torchci/lib/bot/greenBot.ts

```diff
@@ -0,0 +1,212 @@
+      } else {
+        await reactOnComment(ctx, "+1");
+        await addLabels(ctx, [LAND_PENDING]);
+      }
+    }
+  });
+
+  app.on(["pull_request.labeled"], async (ctx) => {
+    if (ctx.payload.label.name === LAND_PENDING) {
+      // Check pull request's head ref and check runs to see if it has all green and merge if it's good
+      // TODO: Add pagination if checks go above 100
+      const payloadData = getPayloadData(ctx);
+      const ref = ctx.payload.pull_request.head.sha;
+      const prNum = ctx.payload.pull_request.number;
+      const checks = await ctx.octokit.checks.listForRef({
+        ...payloadData,
+        ref,
+        per_page: 100,
+      });
+
+      const [successfulChecks, failedChecks] = categorizeChecks(
+        checks.data.check_runs
+      );
+      // If all jobs are successful or neutral, then try merge, else do nothing
+      if (successfulChecks.length === checks.data.check_runs.length) {
+        await tryLandPR({
+          octokit: ctx.octokit,
+          ...payloadData,
+          issue_number: prNum,
+          comment: "All checks passed. Attempting to merge.",
+        });
+      }
+      // If any failed, alert the author
+      else if (failedChecks.length > 0) {
+        await failPR({
+          octokit: ctx.octokit,
+          ...payloadData,
+          issue_number: prNum,
+          comment: "Failed to land PR due to red signal.",
+        });
+      }
+    }
+  });
+
+  app.on("check_run.completed", async (ctx) => {
+    // Check that there are associated PRs that have the label
+    const prWithLabelsAwaitable = ctx.payload.check_run.pull_requests.map(
+      (pr) => {
+        return ctx.octokit.pulls.get({
+          ...getPayloadData(ctx),
+          pull_number: pr.number,
+        });
+      }
+    );
+    const prs = await Promise.all(prWithLabelsAwaitable);
+    const landPendingPrs = prs.filter((pr) => {
+      const ind = pr.data.labels.find((label) => {
+        return label.name === LAND_PENDING;
+      });
+      return ind != null;
+    });
+
+    if (landPendingPrs.length === 0) {
+      return;
+    }
+
+    if (landPendingPrs.length > 1) {
+      // If there's more than 1 land pending PR with the same checks,
+      // then we don't know which one to land
+      return;
+    }
+
+    const landPendingPr = prs[0];
+    // Check the other land signals
+    const checks = await ctx.octokit.checks.listForRef({
+      ...getPayloadData(ctx),
+      ref: landPendingPr.data.head.ref,
+      per_page: 100,
+    });
+
+    const [successfulChecks, failedChecks] = categorizeChecks(
+      checks.data.check_runs
+    );
+
+    const conclusion = ctx.payload["check_run"].conclusion;
+
+    const prData = {
+      octokit: ctx.octokit,
+      ...getPayloadData(ctx),
+      issue_number: landPendingPr.data.number,
+    };
+    if (conclusion !== "success" && conclusion !== "neutral") {
+      // Only alert if this is the first failure
+      if (failedChecks.length < 1) {
```

Comment:
This can can easily race. If two things fail close together (say, someone put up a PR that trivially failed to build), the API call on L98 can return multiple failed checks.

You're probably better off checking the label, which will explicitly tell you whether we've processed a failure before.

Reply:
Also, I think this was intended to be <= 1? Otherwise you're just checking if the length is empty, which seems impossible. Maybe I'm missing smth

Reply:
Hmm that's a good point. WIll use label instead.

Reply:
Is this actually resolved/are you planning on using label? (I would agree that labeling is the best here)